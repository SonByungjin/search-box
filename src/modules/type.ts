export interface CommonError {
  code: number;
  message: string;
  data?: string;
}
