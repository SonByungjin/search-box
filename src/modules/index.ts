import { combineReducers } from "redux";
import { all, call } from "redux-saga/effects";

import { ItemListReducer, ItemListState } from "./itemList";

import { DetailReducer, DetailState } from "./detail/reducer";
import { rootDetailSaga } from "./detail/sagas";

import { SearchReducer, SearchState } from "./search";
import { rootSearchSaga } from "./search/sagas";

export interface RootState {
  itemList: ItemListState;
  search: SearchState;
  detail: DetailState;
}

export const rootReducer = combineReducers({
  itemList: ItemListReducer,
  search: SearchReducer,
  detail: DetailReducer,
});

export function* rootSaga() {
  yield all([
    call(rootSearchSaga),
    call(rootDetailSaga),
  ]);
}
