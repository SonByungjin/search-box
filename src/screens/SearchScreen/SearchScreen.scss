.pg_search {
  &__wrapper {
    width: calc(100vw - 10%);
    height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
    padding: 0 5%;
    > div {
      max-width: 800px;
      width: 100%;
    }
  }

  &__input {
    margin-top: 5%;
    > input {
      width: calc(100% - 10px);
      height: 50px;
      font-size: 18px;
      border: 1px solid #999999;
      padding: 0 0 0 10px;
    }
  }

  &__auto {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: flex-start;
    max-height: 135px;
    overflow-x: hidden;
    overflow-y: scroll;
    padding: 0;

    &::-webkit-scrollbar {
      width: 10px;
      padding: 0;
    }

    &::-webkit-scrollbar-thumb {
      background: #999999;
      border-radius: 8px;
    }

    &__header {
      display: flex;
      align-items: center;
      justify-content: space-evenly;
      width: 100%;
      padding: 5px 0;
      border: 1px solid black;
      background-color: #000000;
      > span {
        color: #ffffff;
      }
      > span:first-child {
        max-width: 400px;
        width: 100%;
      }
    }

    &__container {
      cursor: pointer;
      width: 100%;
      max-height: 50px;
      min-height: 50px;
      display: flex;
      align-items: center;
      justify-content: space-evenly;
      border: 1px solid #999999;
      border-top: none;
      &:hover {
        background: #c8c8c8;
      }
      > span:first-child {
        max-width: 52%;
        width: 100%;
      }
    }
  }
}
