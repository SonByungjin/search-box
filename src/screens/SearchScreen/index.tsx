import { selectItemList } from "modules/itemList";
import { FC, useCallback, useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";

import "./SearchScreen.scss";

interface AutoCompleteItem {
  id: string;
  name: string;
  category: string;
  word: string;
}

const SearchScreen: FC = () => {
  const navigate = useNavigate();
  const itemList = useSelector(selectItemList);
  const [input, setInput] = useState<string>("");
  const [autoCompleteItem, setAutoCompleteItem] = useState<AutoCompleteItem[]>(
    []
  );

  const createAutoCompleteItems = useCallback(() => {
    const autoCompleteItems: AutoCompleteItem[] = [];
    itemList.forEach(({ id, name, category, searchable, count }) => {
      searchable.forEach((word) => {
        if (word.includes(input)) {
          autoCompleteItems.push({
            id,
            name,
            category,
            word,
          });
        }
      });
    });
    return autoCompleteItems;
  }, [input, itemList]);

  const handleItemClick = (id: string, word: string) => {
    navigate(`/detail/id=${id}&word=${word}&input=${input}`);
  };

  useEffect(() => {
    setAutoCompleteItem(input.length >= 2 ? createAutoCompleteItems() : []);
  }, [input, createAutoCompleteItems]);

  return (
    <div className="pg_search__wrapper">
      <div className="pg_search__input">
        <input
          placeholder="검색어 입력"
          value={input}
          onChange={({ target: { value } }) => {
            setInput(value || "");
          }}
        />
      </div>
      {autoCompleteItem.length ? (
        <div className="pg_search__auto">
          <div className="pg_search__auto__header">
            <span>검색어</span>
            <span>id</span>
            <span>name</span>
            <span>category</span>
          </div>
          {autoCompleteItem.map(({ id, name, category, word }, idx) => {
            return (
              <div
                key={idx}
                className="pg_search__auto__container"
                onClick={() => handleItemClick(id, word)}
              >
                <span>{word}</span>
                <span>{id}</span>
                <span>{name}</span>
                <span>{category}</span>
              </div>
            );
          })}
        </div>
      ) : null}
    </div>
  );
};

export { SearchScreen };
