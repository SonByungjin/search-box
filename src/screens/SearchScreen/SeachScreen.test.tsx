import * as React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";
import renderer from "react-test-renderer";

import { rootReducer } from "../../modules";
import { SearchScreen } from "../SearchScreen";

const store = createStore(rootReducer);

describe("SearchScreen", () => {
  test("show render", () => {
    const view = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <SearchScreen />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(view).toMatchSnapshot();
  });
});
