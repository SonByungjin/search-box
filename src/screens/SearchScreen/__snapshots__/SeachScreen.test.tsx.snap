// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen show render 1`] = `
<div
  className="pg_search__wrapper"
>
  <div
    className="pg_search__input"
  >
    <input
      onChange={[Function]}
      placeholder="검색어 입력"
      value=""
    />
  </div>
</div>
`;
